File: examples/java/complex/SecurityUtils.java
package com.bank.api;

import java.security.*;
import java.util.*;
import javax.crypto.*;
import javax.crypto.spec.*;
import java.nio.charset.StandardCharsets;

public class SecurityUtils {

    private static final String HMAC_ALGORITHM = "HmacSHA256";
    private static final String AES_ALGORITHM = "AES/CBC/PKCS5Padding";

    public static String escapeHtml(String input) {
        if (input == null) return "";
        return input
            .replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#x27;");
    }

    public static String hashPassword(String password) throws Exception {
        byte[] salt = new byte[16];
        SecureRandom random = new SecureRandom();
        random.nextBytes(salt);
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        md.update(salt);
        byte[] hash = md.digest(password.getBytes(StandardCharsets.UTF_8));
        return Base64.getEncoder().encodeToString(salt) + ":" + Base64.getEncoder().encodeToString(hash);
    }

    public static boolean verifyPassword(String password, String stored) throws Exception {
        String[] parts = stored.split(":");
        byte[] salt = Base64.getDecoder().decode(parts[0]);
        byte[] storedHash = Base64.getDecoder().decode(parts[1]);
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        md.update(salt);
        byte[] computedHash = md.digest(password.getBytes(StandardCharsets.UTF_8));
        return MessageDigest.isEqual(storedHash, computedHash);
    }

    public static String generateToken() {
        byte[] bytes = new byte[32];
        new SecureRandom().nextBytes(bytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
    }

    public static String hmacSign(String data, String secret) throws Exception {
        Mac mac = Mac.getInstance(HMAC_ALGORITHM);
        SecretKeySpec keySpec = new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), HMAC_ALGORITHM);
        mac.init(keySpec);
        byte[] signature = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
        return Base64.getEncoder().encodeToString(signature);
    }

    public static boolean hmacVerify(String data, String signature, String secret) throws Exception {
        String computed = hmacSign(data, secret);
        return MessageDigest.isEqual(computed.getBytes(), signature.getBytes());
    }

    public static byte[] encrypt(String plaintext, String key) throws Exception {
        byte[] keyBytes = key.getBytes(StandardCharsets.UTF_8);
        SecretKeySpec secretKey = new SecretKeySpec(Arrays.copyOf(keyBytes, 16), "AES");
        byte[] iv = new byte[16];
        new SecureRandom().nextBytes(iv);
        IvParameterSpec ivSpec = new IvParameterSpec(iv);
        Cipher cipher = Cipher.getInstance(AES_ALGORITHM);
        cipher.init(Cipher.ENCRYPT_MODE, secretKey, ivSpec);
        byte[] encrypted = cipher.doFinal(plaintext.getBytes(StandardCharsets.UTF_8));
        byte[] result = new byte[iv.length + encrypted.length];
        System.arraycopy(iv, 0, result, 0, iv.length);
        System.arraycopy(encrypted, 0, result, iv.length, encrypted.length);
        return result;
    }

    public static String decrypt(byte[] ciphertext, String key) throws Exception {
        byte[] keyBytes = key.getBytes(StandardCharsets.UTF_8);
        SecretKeySpec secretKey = new SecretKeySpec(Arrays.copyOf(keyBytes, 16), "AES");
        byte[] iv = Arrays.copyOfRange(ciphertext, 0, 16);
        byte[] encrypted = Arrays.copyOfRange(ciphertext, 16, ciphertext.length);
        IvParameterSpec ivSpec = new IvParameterSpec(iv);
        Cipher cipher = Cipher.getInstance(AES_ALGORITHM);
        cipher.init(Cipher.DECRYPT_MODE, secretKey, ivSpec);
        return new String(cipher.doFinal(encrypted), StandardCharsets.UTF_8);
    }

    public static String sanitizeSqlInput(String input) {
        if (input == null) return "";
        return input.replace("'", "''").replace("\\", "\\\\");
    }

    public static boolean isValidEmail(String email) {
        return email != null && email.matches("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$");
    }

    public static boolean isValidAccountId(String id) {
        return id != null && id.matches("^[A-Za-z0-9-]+$");
    }
}

File: test-fixtures/languages/kotlin/complex/SecurityUtils.kt
package com.bank.app

import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

class SecurityUtils {
    companion object {
        fun escapeHtml(input: String): String {
            return input
                .replace("&", "&amp;")
                .replace("<", "&lt;")
                .replace(">", "&gt;")
                .replace("\"", "&quot;")
                .replace("'", "&#x27;")
        }

        fun isValidAccountId(id: String): Boolean {
            return id.matches(Regex("^[A-Za-z0-9-]+$"))
        }

        fun isValidEmail(email: String): Boolean {
            return email.matches(Regex("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$"))
        }

        fun hashPassword(password: String): String {
            val salt = ByteArray(16)
            SecureRandom().nextBytes(salt)
            val md = MessageDigest.getInstance("SHA-256")
            md.update(salt)
            val hash = md.digest(password.toByteArray())
            return Base64.getEncoder().encodeToString(salt) + ":" + Base64.getEncoder().encodeToString(hash)
        }

        fun verifyPassword(password: String, stored: String): Boolean {
            val parts = stored.split(":")
            if (parts.size != 2) return false
            val salt = Base64.getDecoder().decode(parts[0])
            val storedHash = Base64.getDecoder().decode(parts[1])
            val md = MessageDigest.getInstance("SHA-256")
            md.update(salt)
            val computedHash = md.digest(password.toByteArray())
            return MessageDigest.isEqual(computedHash, storedHash)
        }

        fun generateToken(): String {
            val bytes = ByteArray(32)
            SecureRandom().nextBytes(bytes)
            return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
        }

        fun hmacSign(data: String, secret: String): String {
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(secret.toByteArray(), "HmacSHA256"))
            return Base64.getEncoder().encodeToString(mac.doFinal(data.toByteArray()))
        }

        fun hmacVerify(data: String, signature: String, secret: String): Boolean {
            val computed = hmacSign(data, secret)
            return MessageDigest.isEqual(computed.toByteArray(), signature.toByteArray())
        }

        fun encrypt(plaintext: String, key: String): ByteArray {
            val keyBytes = key.toByteArray().copyOf(16)
            val secretKey = SecretKeySpec(keyBytes, "AES")
            val iv = ByteArray(16)
            SecureRandom().nextBytes(iv)
            val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, IvParameterSpec(iv))
            val encrypted = cipher.doFinal(plaintext.toByteArray())
            return iv + encrypted
        }

        fun sanitizeFilename(filename: String): String {
            return filename.replace(Regex("[^a-zA-Z0-9._-]"), "")
        }

        fun sanitizeLogEntry(entry: String): String {
            return entry.replace("\n", " ").replace("\r", " ")
        }
    }
}
